### PR TITLE
Feat: add support for multi-systems

### PR DESCRIPTION
**Pull request by GaetanLepage**
This PR addresses #400 by allowing to evaluate a PR/revision for multiple systems at once.
You can now provide the `--systems` option with a list of systems to evaluate:
- `--systems current` (default): legacy behavior
- `--systems "x86_64-linux aarch64-darwin"`: evaluate for both intel linux and ARM darwin
- `--systems all`: evaluate on all four main platforms

### Progress

- [x] Overall refactoring, addition of the necessary abstractions
- [x] Adapted all "output" features (`report.json`, `report.md`, stdout...)
- [x] Support for `--eval ofborg`
- [x] Support for `--eval local` (sequential)
- [ ] ~~Support for `--eval local` (optionally parallel)~~ -> Let's do this in a following PR. This one is already big enough.
- [x] Flesh out documentation
- [x] Final code polishing

> [!Note]
>
> This is still a WIP. ~~The tool works fine when using ofborg eval but I have not yet adapted to local evaluation.~~
> The code is quite dirty for now and will require more polish ~~once everything will be working~~.
> Also, I am fully open to suggestions regarding the solution and coding style.

I am eager to receive some feedback !

### Technical details

- Basically, adding this feature requires a pretty heavy refactoring.
Every place where there was a list of attributes/packages/reports, we now need a `dict` mapping each `system` to this list.
- I also needed to change the code for the nix shell to build as it now combines the packages to build for all platforms in the final attribute list.
- The current `--system` option has been softly deprecated (warning + forwarding the value to `--systems`) to avoid breaking the workflow of existing users.
- **Note:** I also had to change the layout of the `report.json`.

cc @Mic92 @natsukium

### CODE REVIEW

**Comment by Mic92** [comment]
Changing report.json will need changes in nipkgs-update as well.

**Comment by Mic92** [comment]
From the code it looks like you handle local evaluation as well, no?

**Comment by GaetanLepage** [comment]
> From the code it looks like you handle local evaluation as well, no?

No, the `build_commit` method of `Review` is broken and has not been adapted.

**Comment by GaetanLepage** [comment]
I have a sequential version working for `--eval local`.
Basically, it calls `nix-env` for each system, one after the other.
We could work on calling them all in parallel, but we would need to provide an option as having 4 of them simultaneously requires a lot of RAM (sometimes even 64GB is not enough).

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review)

- Command: `nixpkgs-review pr 342734`
- Systems: `aarch64-linux`,`x86_64-darwin`,`aarch64-darwin`,`x86_64-linux`

### `aarch64-linux`
<details>
  <summary>:fast_forward: 2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.bambi</li>
    <li>python311Packages.bambi.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 14 packages built:</summary>
  <ul>
    <li>python311Packages.arviz</li>
    <li>python311Packages.arviz.dist</li>
    <li>python311Packages.corner</li>
    <li>python311Packages.corner.dist</li>
    <li>python311Packages.numpyro</li>
    <li>python311Packages.numpyro.dist</li>
    <li>python311Packages.pymc</li>
    <li>python311Packages.pymc.dist</li>
    <li>python312Packages.arviz</li>
    <li>python312Packages.arviz.dist</li>
    <li>python312Packages.corner</li>
    <li>python312Packages.corner.dist</li>
    <li>python312Packages.numpyro</li>
    <li>python312Packages.numpyro.dist</li>
  </ul>
</details>

### `x86_64-darwin`
<details>
  <summary>:fast_forward: 16 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.arviz</li>
    <li>python311Packages.arviz.dist</li>
    <li>python311Packages.bambi</li>
    <li>python311Packages.bambi.dist</li>
    <li>python311Packages.corner</li>
    <li>python311Packages.corner.dist</li>
    <li>python311Packages.numpyro</li>
    <li>python311Packages.numpyro.dist</li>
    <li>python311Packages.pymc</li>
    <li>python311Packages.pymc.dist</li>
    <li>python312Packages.arviz</li>
    <li>python312Packages.arviz.dist</li>
    <li>python312Packages.corner</li>
    <li>python312Packages.corner.dist</li>
    <li>python312Packages.numpyro</li>
    <li>python312Packages.numpyro.dist</li>
  </ul>
</details>

### `aarch64-darwin`
<details>
  <summary>:fast_forward: 6 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.bambi</li>
    <li>python311Packages.bambi.dist</li>
    <li>python311Packages.flowmc</li>
    <li>python311Packages.flowmc.dist</li>
    <li>python311Packages.pymc</li>
    <li>python311Packages.pymc.dist</li>
  </ul>
</details>
<details>
  <summary>:x: 8 packages failed to build:</summary>
  <ul>
    <li>python311Packages.arviz</li>
    <li>python311Packages.arviz.dist</li>
    <li>python311Packages.corner</li>
    <li>python311Packages.corner.dist</li>
    <li>python312Packages.arviz</li>
    <li>python312Packages.arviz.dist</li>
    <li>python312Packages.corner</li>
    <li>python312Packages.corner.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.numpyro</li>
    <li>python311Packages.numpyro.dist</li>
    <li>python312Packages.numpyro</li>
    <li>python312Packages.numpyro.dist</li>
  </ul>
</details>

### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.bambi</li>
    <li>python311Packages.bambi.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 16 packages built:</summary>
  <ul>
    <li>python311Packages.arviz</li>
    <li>python311Packages.arviz.dist</li>
    <li>python311Packages.corner</li>
    <li>python311Packages.corner.dist</li>
    <li>python311Packages.flowmc</li>
    <li>python311Packages.flowmc.dist</li>
    <li>python311Packages.numpyro</li>
    <li>python311Packages.numpyro.dist</li>
    <li>python311Packages.pymc</li>
    <li>python311Packages.pymc.dist</li>
    <li>python312Packages.arviz</li>
    <li>python312Packages.arviz.dist</li>
    <li>python312Packages.corner</li>
    <li>python312Packages.corner.dist</li>
    <li>python312Packages.numpyro</li>
    <li>python312Packages.numpyro.dist</li>
  </ul>
</details>

File: nixpkgs_review/nix.py [hunk 1]
```diff
@@ -305,6 +313,8 @@ def nix_build(
         else "--no-allow-import-from-derivation",
     ]
 
+    # TODO what should we do if we build for both darwin and linux ?
+    # Should this only be affected by the host ?
     if platform == "linux":
```
Comment (GaetanLepage) [inline nixpkgs_review/nix.py:318]: This should remain untouched according to me.

File: nixpkgs_review/review.py [hunk 2]
```diff
@@ -152,33 +175,61 @@ def apply_unstaged(self, staged: bool = False) -> None:
...
-            self.system,
-            self.allow,
-        )
+        # TODO: parallelize ?
+        # parallel version: returning a dict[System, list[Package]]
+        # base_packages = list_packages(
+        #     self.builddir.nix_path,
+        #     self.systems,
+        #     self.allow,
+        # )
```
Comment (GaetanLepage) [inline nixpkgs_review/review.py:190]: We should leave this to another PR.

File: nixpkgs_review/nix.py (hunk 1 above)
Comment (Mic92) [reply nixpkgs_review/nix.py:318]: It depends on the host running the build, not remote builders.

File: nixpkgs_review/nix.py [hunk 3]
```diff
@@ -305,6 +313,8 @@ def nix_build(
         else "--no-allow-import-from-derivation",
     ]
 
+    # TODO what should we do if we build for both darwin and linux ?
+    # Should this only be affected by the host ?
```
Comment (Mic92) [inline nixpkgs_review/nix.py:317]:
```suggestion
```

File: nixpkgs_review/report.py [hunk 4]
```diff
@@ -55,84 +55,129 @@ def ensure(self) -> Path:
...
+        }
+
+
+def order_reports(reports: dict[System, SystemReport]) -> dict[System, SystemReport]:
+    """Ensure that systems are always ordered consistently in reports"""
+    system_order: dict[System, int] = {
+        "x86_64-linux": 0,
+        "aarch64-linux": 1,
+        "x86_64-darwin": 2,
+        "aarch64-darwin": 3,
```
Comment (Mic92) [inline nixpkgs_review/report.py:145]: Could reverse the string here instead and use it as a sort key?

File: nixpkgs_review/review.py [hunk 5]
```diff
@@ -152,33 +175,61 @@ def apply_unstaged(self, staged: bool = False) -> None:
...
         else:
             self.git_merge(reviewed_commit)
 
-        merged_packages = list_packages(
-            self.builddir.nix_path,
-            self.system,
-            self.allow,
-            check_meta=True,
-        )
+        # TODO: parallelize ?
```
Comment (Mic92) [inline nixpkgs_review/review.py:205]:
```suggestion
        # TODO: nix-eval-jobs ?
```

File: nixpkgs_review/report.py [hunk 6]
```diff
@@ -55,84 +55,129 @@ def ensure(self) -> Path:
...
+    }
+    return dict(
+        sorted(
+            reports.items(),
+            key=lambda item: system_order.get(item[0], 100),
```
[the remark this answers is not in the document]
Comment (GaetanLepage) [reply nixpkgs_review/report.py]: This does not quite give the natural order:
- x86_64-darwin
- aarch64-darwin
- x86_64-linux
- aarch64-linux

[the remark this answers is not in the document]
Comment (Mic92) [reply nixpkgs_review/report.py]: It's stable at least :)

File: nixpkgs_review/report.py [hunk 7]
```diff
@@ -55,84 +55,134 @@ def ensure(self) -> Path:
...
+    }
+
+    def sort_func(system: System) -> str:
+        arch, os = system.split("-")
+        return f"{os}-{arch}"
+
+    return dict(
+        sorted(
+            reports.items(),
+            key=lambda item: "".join(reversed(item[0])),
```
Comment (Mic92) [inline nixpkgs_review/report.py:155]:
```suggestion
            key=lambda item: reversed(item[0].split("-")),
```

File: nixpkgs_review/report.py [hunk 8]
```diff
@@ -55,84 +55,134 @@ def ensure(self) -> Path:
...
+    system_order: dict[System, int] = {
+        "x86_64-linux": 0,
+        "aarch64-linux": 1,
+        "x86_64-darwin": 2,
+        "aarch64-darwin": 3,
+    }
+
+    def sort_func(system: System) -> str:
+        arch, os = system.split("-")
+        return f"{os}-{arch}"
```
Comment (Mic92) [inline nixpkgs_review/report.py:150]:
```suggestion
```

File: nixpkgs_review/report.py (hunk 7 above)
Comment (Mic92) [reply nixpkgs_review/report.py:155]: ^ This should not sort architectures more naturally.

Comment (GaetanLepage) [reply nixpkgs_review/report.py]: Ok, adding `reverse=True` I was able to achieve the desired order without needing a static table.

File: README.md [hunk 9]
```diff
@@ -326,10 +345,15 @@ subcommand.
...
-The `--system` flag allows setting a system different from the current one. Note
-that the result nix-shell may not be able to execute all hooks correctly since
+The `--systems` flag allows setting a system different from the current one.
+Note that the result nix-shell may not be able to execute all hooks correctly since
 the architecture/operating system mismatches.
 
+By default, `nixpkgs-review` targets only the current system (`--systems current`).
+You can also explicitly provide one or several systems to target (`--systems "x86_64-linux aarch64-darwin"`).
+The `--systems all` value will build for the four major platforms supported by hydra (`--x86_64-linux`, `aarch64-linux`, `x86_64-darwin` and `aarch64-darwin`).
+Ensure that your system is capable of building for the specified architectures, either locally or through the remote builder protocol.
```
Comment (PerchunPak) [inline README.md:355]: Could you add please a guide/link to a guide for how you build darwin programs on linux?

Comment (PerchunPak) [reply README.md:355]: I think this is it

https://nix.dev/manual/nix/2.18/advanced-topics/distributed-builds

Comment (GaetanLepage) [reply README.md:355]: To my knowledge, there isn't anything special with darwin compared to other foreign architectures.
You simply need to have remote builders able to build darwin packages (a mac basically).
Then, you can `nix-build` everything you want.

Comment (PerchunPak) [reply README.md:355]: Yeah, I mean any foreign architecture. Just darwin is my use case

Comment (GaetanLepage) [reply README.md:355]: Ok, then the page you linked is all you need.
